### PR TITLE
DOC: Fix solve_ivp doc and add dense_output example

### LINKED ISSUE #9667: DOC: scipy.integrate.solve_ivp example
This example looks wrong.  Shouldn't it be
> def hit_ground(t, y): return y[0]
if you want to stop when position is 0?

Also, I get this warning from this example:
> /usr/local/lib/python3.4/dist-packages/scipy/integrate/_ivp/rk.py:145: RuntimeWarning: divide by zero encountered in double_scalars  max(1, SAFETY * error_norm ** (-1 / (order + 1))))

>Cannon fired upward with terminal event upon impact. The terminal and direction fields of an event are >applied by monkey patching a function. Here y[0] is position and y[1] is velocity. The projectile starts at >position 0 with velocity +10. Note that the integration never reaches t=100 because the event is terminal.
>>> def upward_cannon(t, y): return [y[1], -0.5]
>>> def hit_ground(t, y): return y[1]

### PR DESCRIPTION
**Pull request by MatthewFlamm**
Fix cannonball example in solve_ivp.  Fixes #9667.  Also extends cannonball example to include usage of dense_output, which is currently missing from examples. 

I also noticed that the input parameter types are inconsistently applied in this docstring, but I'm not sure what the preferred format is.  I see lines like  `dense_output : bool, optional`, but other lines are like `t_eval : array_like or None, optional`.  It is also unclear to me when `type1, type2, or type3` is preferred or `{type1, type2, type3}`, both of which appear in this docstring. The numpy docstring guide isn't very clear on the preferred style, does it not matter?  Please let me know if this should be in a different issue/PR or just ignored.

### CODE REVIEW

**Comment by Kai-Striega** [comment]
@MatthewFlamm the failing test wasn't related to your changes and have been fixed elsewhere. I've re-run the failing test and looked over your changes - they all look good. If the tests pass I'll have double check the changes and (hopefully) merge.

**Comment by MatthewFlamm** [comment]
Thanks @Kai-Striega.  This is a recurring theme as I think 3/5 or more of my PRs have had failing tests unrelated to the PR.  This must be a drag on your and other dev's time. Thank you for your efforts!

File: scipy/integrate/_ivp/ivp.py [hunk 1]
```diff
@@ -412,15 +412,32 @@ def solve_ivp(fun, t_span, y0, method='RK45', t_eval=None, dense_output=False,
...
+    [array([40.])]
     >>> print(sol.t)
     [0.00000000e+00 9.99900010e-05 1.09989001e-03 1.10988901e-02
-     1.11088891e-01 1.11098890e+00 1.11099890e+01 2.00000000e+01]
+     1.11088891e-01 1.11098890e+00 1.11099890e+01 4.00000000e+01]
+
+    Use dense_output and events to find position, which is 100, at apex of
+    trajectory of cannonball. Apex is not defined as terminal, so both apex
+    and hit_ground are found. There is no information at t=20, so the sol
+    attribute is used to evaluate the solution.  The sol attribute is
```
Comment (Kai-Striega) [inline scipy/integrate/_ivp/ivp.py]: Sometimes you have two spaces between sentences, when only one is needed.

File: scipy/integrate/_ivp/ivp.py [hunk 2]
```diff
@@ -412,15 +412,32 @@ def solve_ivp(fun, t_span, y0, method='RK45', t_eval=None, dense_output=False,
...
     >>> sol = solve_ivp(upward_cannon, [0, 100], [0, 10], events=hit_ground)
     >>> print(sol.t_events)
-    [array([ 20.])]
+    [array([40.])]
     >>> print(sol.t)
     [0.00000000e+00 9.99900010e-05 1.09989001e-03 1.10988901e-02
-     1.11088891e-01 1.11098890e+00 1.11099890e+01 2.00000000e+01]
+     1.11088891e-01 1.11098890e+00 1.11099890e+01 4.00000000e+01]
+
+    Use dense_output and events to find position, which is 100, at apex of
```
[the remark this answers is not in the document]
Comment (Kai-Striega) [reply scipy/integrate/_ivp/ivp.py]: On an extra note, variable names should be written between single backtick. The variable `events` is also documented as `t_events`.

**Comment by Kai-Striega** [comment]
Regarding your docstring questions NumPy provides a  [guide](https://numpydoc.readthedocs.io/en/latest/format.html#numpydoc-docstring-guide)  which includes an [example](https://numpydoc.readthedocs.io/en/latest/example.html#example) to look at.

``dense_output : bool, optional`` vs ``t_eval : array_like or None, optional``
It will depend on the input your expecting. The variable `dense_output` must be either yes or no - the output is either dense or not dense. Passing `None` doesn't make sense and should not be input.

`t_eval` is more complicated. It could be an [array_like](https://numpydoc.readthedocs.io/en/latest/format.html#id16) object that contains the times you want, or, if you don't want any, an empty `array_like` object. However, a mutable default argument can lead to [unexpected behaviour](https://docs.python-guide.org/writing/gotchas/). The Pythonic solution is to pass `None` by default, which the user may also set explicitly and is a valid argument type.

``type1, type2, or type3`` vs ``{type1, type2, or type3}`` should be the first. The use of curly braces is described in the docs guide as:

> When a parameter can only assume one of a fixed set of values, those values can be listed in braces, with the default appearing first: 

    order : {'C', 'F', 'A'}
        Description of `order`.

File: scipy/integrate/_ivp/ivp.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MatthewFlamm) [reply scipy/integrate/_ivp/ivp.py]: Your version reads much better.

> On an extra note, variable names should be written between single backtick. The variable `events` is also documented as `t_events`.

I'm not sure what your comment is here about `events` and `t_events`.  The latter is returned when the former is not None.

**Comment by MatthewFlamm** [comment]
Made a quick clean up of the type issues.  I don't like how the default values are contingent on whether a solver supports it coupled with the solver support is inconsistently defined in each case.  I'd propose something like, but this seems best as a separate discussion.

```
arg1 : float, optional
     Supported solvers: Solver1, Solver2
     Rest of docstring
```

**Comment by Kai-Striega** [comment]
@rgommers can we merge this? I'm not sure why the test is failing. It's unrelated to the changes made here and occured in #9676.

**Comment by MatthewFlamm** [comment]
Now that this PR has grown a bit, does it make sense to include the change from #9694 here?

**Comment by Kai-Striega** [comment]
I think to leave the two PRs apart is the better option. I've had a quick look over it, they look like the error is the same. Maybe there's something occuring in only your build?

@mdhaber do you know how you managed to fix the error when it occurred in #9676?

**Comment by mdhaber** [comment]
Rebase on master, build (maybe?) and force push.

**Comment by MatthewFlamm** [comment]
In my limited understanding of how this works, CircleCI doesn't merge the PR into master (in this case) for the test, but Travis CI does.  I wonder if one of the workarounds on the link here would align the testing for SciPy among CircleCI and Travis CI.

https://discuss.circleci.com/t/show-test-results-for-prospective-merge-of-a-github-pr/1662

I can try to rebase and force push.  If this doesn't work, I haven't been previously successful at getting a development build environment setup correctly, but it would be worth another try.

**Comment by MatthewFlamm** [comment]
CircleCI pypy3 test passed.  I learned about how to use rebase and force push in this PR too, Thanks all!

**Comment by Kai-Striega** [comment]
I'm happy to merge this. I've had another look over the changes - they all look good and all test are (finally) passing. I'll review #9694 today.

Thank you for taking the time to contribute.

**Pull request merged**